[Dat][P3] Implement Data Transform Component

deerhunter4 opened the pull request:
This pull request combines three components that are three separate tickets in Trello and should be pull requested separately. However, we are short on time so I combined them.

[Dat][P3] Implement Data Transform Component
[Dat][P3] Implement File Writer Component
[Dat][P4] Implement Plot Writer Component

deerhunter4 committed:
[Dat][P3] Implement Data Transform Component

[Dat][P3] Implement File Writer Component.
Two new files were created data_transformer.py and file_writer.py.
Code that use two above files wasa added to main.py.
In api_response_parser.py longitude and latitude were
removed from parse_openmeteo_response output.

deerhunter4 committed:
.gitignore updated

CSV files are ignored

deerhunter4 committed:
[Dat][P4] Implement Plot Writer Component

New component responsible for drawing and saving plot
was created: plot.py.
main.py was updated to use plot.py
.gitignore was updated - ignore PDF files

deerhunter4 committed:
[Dat][P4] Plot Writer Component Update

Grid added to the plot and plot size increased.

KrzysiekJodlowski commented on .gitignore:
```diff
@@ -2,4 +2,6 @@
\ No newline at end of file
+*.txt
+*.csv
+*.pdf
```
There is no newline at the end of the file.

deerhunter4 replied:
Done.

KrzysiekJodlowski commented on data_transformer.py:
```diff
@@ -0,0 +1,26 @@
+
+def create_dataframe(dict1, dict2, dict3):
+    all_data_dict = {**dict1, **dict2, **dict3}
+    df = pd.DataFrame(all_data_dict)
```
I wouldn't say I like the way this var is named, also we already discussed that a couple of times. Can we name it the right way?

deerhunter4 replied:
Done. I changed "df" to "full_forecast_dataframe".

KrzysiekJodlowski commented on data_transformer.py:
```diff
@@ -0,0 +1,26 @@
+    df = pd.DataFrame(all_data_dict)
+
+    # check if for each forecast point the time period (date and hour) is the same and create variables date and hour
+    if df['hours_meteostat'][0][:10] == df['hours_openmeteo'][0][:10] and df['hours_meteostat'][0][:10] == df['hours_weatherapi'][0][:10]:
+        df['date'] = df['hours_meteostat'][0][:10]
+
+    df.insert(loc=1, column='hour', value="empty")
+    for row_no in range(0, len(df)):
+        if df['hours_meteostat'][row_no][11:16] == df['hours_openmeteo'][row_no][11:16] == df['hours_weatherapi'][row_no][11:16]:
+            df.loc[row_no, 'hour'] = df['hours_meteostat'][row_no][11:16]
```
These vars here used in the `if` condition and the `for` loop look too ambiguous (while looking at them I don't know what they are and what they represent). Can you please extract them to properly named local vars so that anyone looking at this code knows what happens here? This also applies to all the "magic numbers" (please check for yourself what it means). And in general, as a reminder, code is to be read by humans;)

KrzysiekJodlowski replied:
One more thing in the spirit of clean code. If your methods are getting longish, and especially if you feel the need to comment some of its part, its becoming a good candidate to extract such code part to a separate method. This way all thing becomes more readable, and have names reflecting what's happening there.

deerhunter4 replied:
@KrzysiekJodlowski 
Done. I made two new functions and gave the vars names. I couldn't change the "magic numbers", but with variable names, they may be clearer now.

KrzysiekJodlowski replied:
@deerhunter4 good job, that's basically what I meant about doing something about the magic numbers, naming them, to give them sense.

KrzysiekJodlowski commented on main.py:
```diff
@@ -30,3 +32,13 @@
     print("\nmeteostat forecast\n")
     print(meteostat_forecast)
+
+    # create dataframe from diccionaries
+    dataframe = create_dataframe(weatherapi_forecast, openmeteo_forecast, meteostat_forecast)
+    print(dataframe)
+
+    # save dataframe as CSV file
+    write_csv(dataframe, parameters)
+
+    # create plot 
+    create_plot(dataframe, parameters)
```
Do you think we really need those comments? These are one-liners and since you named the methods here in a readable way I think it's already self-explanatory.

deerhunter4 replied:
Done.

KrzysiekJodlowski commented on plot.py:
```diff
@@ -0,0 +1,37 @@
+import matplotlib.pyplot as plt
+import pandas as pd
+
+# df = pd.read_csv("test_for_plot.txt", sep="\t")
+# df['mean_rain'] = [0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 11.0, 20.0, 17.0,
+#                    6.0, 0.5, 0.2, 0.2, 0.0, 0.0, 0.0, 0.0, 0.0,
+#                    0.0, 0.0, 0.0, 0.2, 0.1, 0.2]
```
Why do we have here a commented code? If it's needed please uncomment it, and if not remove it.

deerhunter4 replied:
Done

KrzysiekJodlowski commented on data_transformer.py:
```diff
@@ -0,0 +1,26 @@
+import pandas as pd
+
+
+def create_dataframe(dict1, dict2, dict3):
```
What are `dict1` etc? Please name those parameters in any meaningful way.

deerhunter4 replied:
Done

KrzysiekJodlowski commented on file_writer.py:
```diff
@@ -0,0 +1,6 @@
+import pandas as pd
+
+
+def write_csv(df, parameters):
```
```suggestion
def write_csv_file(data_frame, parameters):
```
I'll leave the method name up to you, but I'd suggest naming the 1st param in a meaningful way.

deerhunter4 replied:
Done.

KrzysiekJodlowski commented on plot.py:
```diff
@@ -0,0 +1,37 @@
+
+def create_plot(df, parameters):
+    # create plot axes limits
+    max_rain = 10
+    rain_max = df["mean_rain"].max()*1.1
+    if max_rain < rain_max:
+        max_rain = rain_max
+
+    max_temp = df["mean_temp"].max()*1.1
+    min_temp = df["mean_temp"].min() - df["mean_temp"].max()*0.1
```
Why/what are `10`, `1.1`, or `0.1`? Below I'm also seeing `12`, `20`, `16`, and `300`. Again, these are "magic numbers". Please handle that in accordance with clean code guides.

deerhunter4 replied:
@KrzysiekJodlowski 
Here I would need a little bit of guidance from you since I will not start to read a book right now. The first set of numbers that you mention is arbitrarily chosen to make the precipitation plot more intuitive. When rain is small or non then scale will keep 10 mm as max and not stretch bars up. Also, max scale precipitation will be always 10% higher than max measurement for clarity so the top of the bars will not touch the top of the plot.
Another set of numbers is responsible for font size or plot size.
What should I do? Explain it in the comments?

KrzysiekJodlowski replied:
@deerhunter4 let me make that easier for you 😉 You can start by looking at this webpage: https://refactoring.guru/pl/replace-magic-number-with-symbolic-constant

deerhunter4 replied:
Done. I created constant values only for the first three "magic numbers', plus I commented it very well. 
Other numbers for me are self-explanatory like 'fontsize' and others needed just a few words of explanation as a commented text near to them.

KrzysiekJodlowski commented on plot.py:
```diff
@@ -0,0 +1,37 @@
+#                    0.0, 0.0, 0.0, 0.2, 0.1, 0.2]
+
+
+def create_plot(df, parameters):
```
What's a `df`? I should know that by just taking a glance at the code I'm reading.

deerhunter4 replied:
Done. Changed to full_forecast_dataframe.

KrzysiekJodlowski commented on plot.py:
```diff
@@ -0,0 +1,37 @@
+
+    # create plot
+    fig, ax = plt.subplots(figsize=(20, 12))
+    # plt.tick_params(axis='y', which='both', labelleft=False, labelright=True)
+    # fig = plt.figure(figsize=(12, 8))
```
Again, commented code, is to be uncommented or removed.

deerhunter4 replied:
Done.

KrzysiekJodlowski commented on plot.py:
```diff
@@ -0,0 +1,37 @@
+    ax2.set_ylabel('Temperature [C]', fontsize=16)
+    ax2.set_ylim(min_temp, max_temp)
+    plot_name = f"Better_weather_{parameters.location}_{parameters.forecast_period}_{parameters.date}.pdf"
+    fig.savefig(plot_name, dpi=300)
+    plt.show()
```
I'd suggest making an empty line before that part. It's responsible for saving/showing the plot file, which differs from its configuration.

deerhunter4 replied:
Done.

KrzysiekJodlowski reviewed (changes requested):
Looks ok, I just left some mainly clean code-related comments. 

Since I see that you still don't have the instinct to write code in a clean way, I highly recommend reading a bit about it before you tackle the next challenges (there are plenty of sources on the internet about it, but I believe the best one is ["Clean Code" book by Robert C. Martin](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882)).

Also, in the PR description you're saying you are merging a few tickets here - can you please list them out in the description so that I know which ones we're talking about? (it'd be best if you craft a habit for doing that always)

And, I also do not see any README file updates which now mention that some features are still under development - I guess some of this info would be outdated if you merge this PR, am I right?

deerhunter4 committed:
[Dat][P3] data_transformer.py update
Two new functions created: check_date and check_hour.
create_dataframe function simplified
Variable names added
.gitignore file corrected

deerhunter4 committed:
[Dat][P3-P4] Files update.

KrzysiekJodlowski commented:
@deerhunter4  just a note, I know that you'll not read the book just like that but:
* You don't have to, you can just pick a chapter
* Remember that the Internet is your friend
* You can also ask any AI tool about it

Don't want to sound grumpy but you may ask why I'm so pushy about such things. As in every profession, any newcomer **considering it for real** just needs to grasp: 
* core basics that in time should become a habit, e.g. for programming (level junior) one of those is clean code basics
* specific flow, or approach of work, in this case, effective searching for the information you need to get is a must

We can talk about it when we go back to our status meetings. I can prepare a list of the core developer basics for you to catch up with them at your pace.

deerhunter4 committed:
[Dat][P4] Implement Plot Writer Component Update

'Magic numbers' became constan values and are explained
in commented text, like some other numbers that did
not get their own constan values.

deerhunter4 committed:
[Dat][P4] Implement Plot Writer Component Update

KrzysiekJodlowski reviewed (approved):
Great job @deerhunter4 👏
And thank you for your patience, believe me or not, that will pay off in the future🙂
